workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/internal/flows/finality/ObserverAwareFinalityFlowHandler.kt

```diff
@@ -1,25 +1,48 @@
 package com.r3.corda.lib.tokens.workflows.internal.flows.finality
 
 import co.paralleluniverse.fibers.Suspendable
+import net.corda.core.crypto.SecureHash
+import net.corda.core.flows.FlowException
 import net.corda.core.flows.FlowLogic
 import net.corda.core.flows.FlowSession
-import net.corda.core.flows.ReceiveFinalityFlow
+import net.corda.core.flows.ReceiveTransactionFlow
+import net.corda.core.identity.Party
 import net.corda.core.node.StatesToRecord
 import net.corda.core.transactions.SignedTransaction
 import net.corda.core.utilities.unwrap
 
-class ObserverAwareFinalityFlowHandler(val otherSession: FlowSession) : FlowLogic<SignedTransaction?>() {
-    @Suspendable
-    override fun call(): SignedTransaction? {
-        val role = otherSession.receive<TransactionRole>().unwrap { it }
-        val statesToRecord = when (role) {
-            TransactionRole.PARTICIPANT -> StatesToRecord.ONLY_RELEVANT
-            TransactionRole.OBSERVER -> StatesToRecord.ALL_VISIBLE
-        }
-        // If states are issued to self, then ReceiveFinalityFlow does not need to be invoked.
-        return if (!serviceHub.myInfo.isLegalIdentity(otherSession.counterparty)) {
-            subFlow(ReceiveFinalityFlow(otherSideSession = otherSession, statesToRecord = statesToRecord))
-        } else null
-    }
-}
+class ObserverAwareFinalityFlowHandler(val otherSession: FlowSession, val expectedTxId: SecureHash? = null) : FlowLogic<SignedTransaction?>() {
+	@Suspendable
+	override fun call(): SignedTransaction? {
+
+		val ourKeys = serviceHub.keyManagementService.filterMyKeys(serviceHub.keyManagementService.keys)
+
+		val role = otherSession.receive<TransactionRole>().unwrap { it }
+		val statesToRecord = role.toStatesToRecord()
+
+		return if (otherSession.counterparty.owningKey in ourKeys) null else {
```

Comment:
I think the original logic that used `serviceHub.myInfo.isLegalIdentity(otherSession.counterparty)` is better and more efficient.

Reply:
Reverted.